build-system/pr-check.js

```diff
@@ -0,0 +1,190 @@
+ * is, commands are executed on after the other (see the exec
+ * function). Should a command fail, this script will then also fail.
+ * This script attempts to introduce some granularity for our
+ * presubmit checking, via the determineBuildTargets method.
+ */
+const child_process = require('child_process');
+const path = require('path');
+
+/**
+ * @param {string} str
+ * @param {string} prefix
+ * @return {boolean}
+ */
+function hasPrefix(str, prefix) {
+  return str.indexOf(prefix) == 0;
+}
+
+/**
+ * @param {string} str
+ * @param {string} prefix
+ * @return {boolean}
+ */
+function hasSuffix(str, suffix) {
+  return str.lastIndexOf(suffix) == str.length - suffix.length;
+}
+
+/**
+ * Executes the provided command, returning its stdout as an array of lines.
+ * This will throw an exception if something goes wrong.
+ * @param {string} cmd
+ * @return {!Array<string>}
+ */
+function exec(cmd) {
+  return child_process.execSync(cmd, {'encoding': 'utf-8'}).trim().split('\n');
+}
+
+/**
+ * Executes the provided command; terminates this program in case of failure.
+ * @param {string} cmd
+ */
+function execOrDie(cmd) {
+  console /*OK*/.log(`\npr-check.js: ${cmd}\n`);
+  const p =
+      child_process.spawnSync('/bin/sh', ['-c', cmd], {'stdio': 'inherit'});
+  if (p.status != 0) {
+    console /*OK*/.error(
+        `\npr-check.js - exiting due to failing command: ${cmd}\n`);
+    process.exit(p.status)
+  }
+}
+
+/**
+ * For a provided commit range identifiying a pull request (PR),
+ * yields the list of files.
+ * @param {string} travisCommitRange
+ * @return {!Array<string>}
+ */
+function filesInPr(travisCommitRange) {
+  return exec(`git diff --name-only ${travisCommitRange}`);
+}
+
+/**
+ * Determines whether the given file belongs to the Validator webui,
+ * that is, the 'VALIDATOR_WEBUI' target.
+ * @param {string} filePath
+ * @return {boolean}
+ */
+function isValidatorWebuiFile(filePath) {
+  return hasPrefix(filePath, 'validator/webui');
+}
+
+/**
+ * Determines whether the given file belongs to the validator,
+ * that is, the 'VALIDATOR' target. This assumes (but does not
+ * check) that the file is not part of 'VALIDATOR_WEBUI'.
+ * @param {string} filePath
+ * @return {boolean}
+ */
+function isValidatorFile(filePath) {
+  if (hasPrefix(filePath, 'validator/')) return true;
+  if (!hasSuffix(path.dirname(filePath), '0.1') &&
+      !hasSuffix(path.dirname(filePath), 'test'))
+    return false;
+  const name = path.basename(filePath);
+  return hasPrefix(name, 'validator-') &&
+      (hasSuffix(name, '.out') || hasSuffix(name, '.html') ||
+       hasSuffix('.protoascii'));
+}
+
+/**
+ * Determines the targets that will be executed by the main method of
+ * this script. The order within this function matters.
+ * @param {!Array<string>} filePaths
+ * @returns {!Set<string>}
+ */
+function determineBuildTargets(filePaths) {
+  const targetSet = new Set();
+  for (p of filePaths) {
```

Comment:
My impression is that since is used inside the Travis setup, the target Node.js is whatever nvm considers stable - as specified by the .travis.yml in the top-level directory. So I installed that ("nvm use stable") and that's what I used for testing / playing around. So I see v6.5.0 at the moment. If we want to support old versions, which ones do we care about?
Reason I used these newer features is that they make the code more readable and less bug prone, like forgetting hasOwnProperty checks and so on.
But I'm OK with backporting if you really think it's better. Thank you.